validator_client/src/http_api/keystores.rs

```diff
@@ -0,0 +1,290 @@
+                            "public_key" => ?public_key,
+                        );
+                    }
+                }
+            }
+
+            validator_store.import_slashing_protection(slashing_protection)
+        } else {
+            warn!(log, "No slashing protection data provided with keystores");
+            Ok(())
+        };
+
+    // Import each keystore. Some keystores may fail to be imported, so we record a status for each.
+    let mut statuses = Vec::with_capacity(request.keystores.len());
+
+    for (KeystoreJsonStr(keystore), password) in request
+        .keystores
+        .into_iter()
+        .zip(request.passwords.into_iter())
+    {
+        let pubkey_str = keystore.pubkey().to_string();
+
+        let status = if let Err(e) = &slashing_protection_status {
+            // Slashing protection import failed, do not attempt to import the key. Record an
+            // error status.
+            Status::error(
+                ImportKeystoreStatus::Error,
+                format!("slashing protection import failed: {:?}", e),
+            )
+        } else if let Some(runtime) = runtime.upgrade() {
+            // Import the keystore.
+            match import_single_keystore(
+                keystore,
+                password,
+                validator_dir.clone(),
+                &validator_store,
+                runtime,
+            ) {
+                Ok(status) => Status::ok(status),
+                Err(e) => {
+                    warn!(
+                        log,
+                        "Error importing keystore, skipped";
+                        "pubkey" => pubkey_str,
+                        "error" => ?e,
+                    );
+                    Status::error(ImportKeystoreStatus::Error, e)
+                }
+            }
+        } else {
+            Status::error(
+                ImportKeystoreStatus::Error,
+                "validator client shutdown".into(),
+            )
+        };
+        statuses.push(status);
+    }
+
+    Ok(ImportKeystoresResponse { data: statuses })
+}
+
+fn import_single_keystore<T: SlotClock + 'static, E: EthSpec>(
+    keystore: Keystore,
+    password: ZeroizeString,
+    validator_dir_path: PathBuf,
+    validator_store: &ValidatorStore<T, E>,
+    runtime: Arc<Runtime>,
+) -> Result<ImportKeystoreStatus, String> {
+    // Check if the validator key already exists, erroring if it is a remote signer validator.
+    let pubkey = keystore
+        .public_key()
+        .ok_or_else(|| format!("invalid pubkey: {}", keystore.pubkey()))?;
+    if let Some(def) = validator_store
+        .initialized_validators()
+        .read()
+        .validator_definitions()
+        .iter()
+        .find(|def| def.voting_public_key == pubkey)
+    {
+        if !def.signing_definition.is_local_keystore() {
+            return Err("cannot import duplicate of existing remote signer validator".into());
```

Comment:
I notice that an `enabled = false` validator might not appear in `/list` but they'll trigger this duplicate warning here. I don't see this as a significant issue and I assume there's going to be some edge-cases when meshing the standard API with our validator management.

Reply:
Yeah good point. I'm OK with the current behaviour, but yeah, might make sense to re-assess in future.